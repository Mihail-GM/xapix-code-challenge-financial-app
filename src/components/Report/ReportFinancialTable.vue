<template>
	<div id="financial-table-container" class="mb-16">
		<div id="financial-table-loading" v-if="actionState === 'LOADING'">
			<v-progress-linear
				indeterminate
				color="green"
			/>

			<div class="d-flex justify-center align-center pa-10">
				Loading... Please wait
			</div>
		</div>

		<v-alert
			v-else-if="actionState === 'ERRORED'"
			id="financial-table-errored"
			dense
			outlined
			type="error"
		>
			There was an error when loading data. Please try again later.
		</v-alert>

		<v-data-table
			v-else
			id="financial-table"
			:headers="headers"
			:items="financial"
			sort-by="primary_contact_email"
			class="elevation-2 px-10 text-xl-h4"
		/>
		<small>*you can sort data by just clicking on the headers</small>
	</div>
</template>

<script>
	import {mapGetters} from "vuex";

	export default {
		name: "ReportFinancialTable",

		methods: {
			async getDataForReport() {
				await this.$store.dispatch('financialReport/fetchCalculatedDataForReport',);
			},
		},

		data() {
			return {
				headers: [
					{text: 'Period', align: 'start', sortable: true, value: 'dataRanges'},
					{text: 'Product Sales', align: 'start', sortable: true, value: 'productSales'},
					{text: 'Shipping Revenue', align: 'start', sortable: true, value: 'shippingRevenue'},
					{text: 'Other Income', align: 'start', sortable: true, value: `otherIncome`},
					{text: 'Total sales', align: 'start', sortable: true, value: `totalSales`},
					{text: 'Cost', align: 'start', sortable: true, value: `cost`},
					{text: 'Total', align: 'start', sortable: true, value: `total`},
				],
			}
		},

		computed: {
			...mapGetters(
				{
					financial: 'financialReport/getCalculatedDataForReport',
					actionState: 'financialReport/getCurrentState'
				}
			)
		},
		created() {
			this.getDataForReport();
		},
	}
</script>

<style scoped>

</style>