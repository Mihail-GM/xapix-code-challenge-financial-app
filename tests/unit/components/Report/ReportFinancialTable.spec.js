import Vuetify from 'vuetify'
import Vuex from "vuex";

import {createLocalVue, mount} from '@vue/test-utils'
import {merge} from "lodash";

import ReportFinancialTable from "@/components/Report/ReportFinancialTable";
import financialReportStore from "@/store/modules/financialReport";

describe('ReportFinancialTable.vue', () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);
	let vuetify;
	let getters;

	beforeEach(() => {
		vuetify = new Vuetify();
	})

	function createStore(overrides) {
		let currentState = "DATA_LOADED";
		let financial = [
				{
					"dataRanges": "1/2021",
					"productSales": 17.24,
					"shippingRevenue": 2.44,
					"otherIncome": 0,
					"cost": -3.32,
					"totalSales": 19.68,
					"total": 16.36
				},
				{
					"dataRanges": "2021",
					"productSales": 21.55,
					"shippingRevenue": 3.05,
					"otherIncome": 0,
					"cost": -4.15,
					"totalSales": 24.6,
					"total": 20.45
				},
				{
					"dataRanges": "2/2021",
					"productSales": 4.31,
					"shippingRevenue": 0.61,
					"otherIncome": 0,
					"cost": -0.83,
					"totalSales": 4.92,
					"total": 4.09
				}
			]
		;

		getters = {
			getCurrentState: () => currentState,
			getCalculatedDataForReport: () => financial
		}

		const defaultStoreConfig = {
			modules: {
				financialReport: {
					namespaced: true,
					state: {
						currentState,
						financial,
					},
					getters: financialReportStore.getters,

				}
			}
		};

		return new Vuex.Store(merge(defaultStoreConfig, overrides));
	}

	const createWrapper = (overrides) => {
		const defaultMountingOptions = {
			localVue,
			vuetify,
			store: createStore(),

			mocks: {
				// Vue Router
				$router: {
					push: () => {
					}
				},
			},
		};

		return mount(ReportFinancialTable, merge(defaultMountingOptions, overrides))
	}

	it('should render component ', () => {
		const wrapper = createWrapper();

		expect(wrapper.is(ReportFinancialTable)).toBe(true);
	})

	it('should match snapshot', () => {
		const wrapper = createWrapper()

		expect(wrapper.html()).toMatchSnapshot()
	})
})
