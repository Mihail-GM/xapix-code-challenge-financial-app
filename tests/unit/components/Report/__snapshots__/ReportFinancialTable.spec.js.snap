// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportFinancialTable.vue should match snapshot 1`] = `
<div id="financial-table-container" class="mb-16">
  <div class="v-data-table elevation-2 px-10 text-xl-h4 v-data-table--has-bottom theme--light" id="financial-table">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Period: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Period</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Product Sales: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Product Sales</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Shipping Revenue: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Shipping Revenue</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Other Income: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Other Income</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Total sales: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Total sales</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Cost: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Cost</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Total: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Total</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr class="v-data-table__empty-wrapper">
            <td colspan="7">No data available</td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-39" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-39" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">â€“</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
    </div>
  </div> <small>*you can sort data by just clicking on the headers</small>
</div>
`;
